src/nvim/os/env.c

```diff
@@ -87,3 +97,598 @@ void os_get_hostname(char *hostname, size_t len)
+/// This is not very memory efficient, this expects the result to be freed
+/// again soon.
+/// @param src String containing environment variables to expand
+/// @see {expand_env}
+char_u *expand_env_save(char_u *src)
+{
+  return expand_env_save_opt(src, false);
+}
+
+/// Similar to expand_env_save() but when "one" is `true` handle the string as
+/// one file name, i.e. only expand "~" at the start.
+/// @param src String containing environment variables to expand
+/// @param one Should treat as only one file name
+/// @see {expand_env}
+char_u *expand_env_save_opt(char_u *src, bool one)
+{
+  char_u *p = xmalloc(MAXPATHL);
+  expand_env_esc(src, p, MAXPATHL, false, one, NULL);
+  return p;
+}
+
+/// Expand environment variable with path name.
+/// "~/" is also expanded, using $HOME. For Unix "~user/" is expanded.
+/// Skips over "\ ", "\~" and "\$" (not for Win32 though).
+/// If anything fails no expansion is done and dst equals src.
+/// @param src Input string e.g. "$HOME/vim.hlp"
+/// @param dst Where to put the result
+/// @param dstlen Maximum length of the result
+void expand_env(char_u *src, char_u *dst, int dstlen)
+{
+  expand_env_esc(src, dst, dstlen, false, false, NULL);
+}
+
+/// Expand environment variable with path name and escaping.
+/// "~/" is also expanded, using $HOME. For Unix "~user/" is expanded.
+/// Skips over "\ ", "\~" and "\$" (not for Win32 though).
+/// If anything fails no expansion is done and dst equals src.
+/// startstr recognize the start of a new name, for '~' expansion.
+/// @param srcp Input string e.g. "$HOME/vim.hlp"
+/// @param dst Where to put the result
+/// @param dstlen Maximum length of the result
+/// @param esc Should we escape spaces in expanded variables?
+/// @param one Should we expand more than one '~'?
+/// @param startstr Common prefix for paths, can be NULL
+void expand_env_esc(char_u *srcp, char_u *dst, int dstlen, bool esc, bool one,
+                    char_u *startstr)
+{
+  char_u      *src;
+  char_u      *tail;
+  int c;
+  char_u      *var;
+  bool copy_char;
+  bool mustfree;                 /* var was allocated, need to free it later */
+  bool at_start = true;           /* at start of a name */
+  int startstr_len = 0;
+
+  if (startstr != NULL)
+    startstr_len = (int)STRLEN(startstr);
+
+  src = skipwhite(srcp);
+  --dstlen;                 /* leave one char space for "\," */
+  while (*src && dstlen > 0) {
+    copy_char = true;
+    if ((*src == '$') || (*src == '~' && at_start)) {
+      mustfree = false;
+
+      /*
+       * The variable name is copied into dst temporarily, because it may
+       * be a string in read-only memory and a NUL needs to be appended.
+       */
+      if (*src != '~') {                                /* environment var */
+        tail = src + 1;
+        var = dst;
+        c = dstlen - 1;
+
+#ifdef UNIX
+        /* Unix has ${var-name} type environment vars */
+        if (*tail == '{' && !vim_isIDc('{')) {
```

Comment:
Check out [`:h 'isident'`](http://vimpluginloader.sourceforge.net/doc/options.txt.html#.27isident.27). You can make any ASCII character be a part of the identifier by using this option.